### PR TITLE
Fix race condition in once

### PR DESCRIPTION
**Pull request by forslund**
This is a naive locking approach to the issue reported in #76

This PR adds a check in the once-wrapper that listener method is still registered before removing. To ensure atomic execution a lock is added around removals of listeners.

### CODE REVIEW

File: pyee/_base.py [hunk 1]
```diff
@@ -131,18 +139,24 @@ def g(*args, **kwargs):
         else:
             return _wrapper(f)
 
+    def _remove_listener(self, event, f):
+        """Naked unprotected removal."""
```
Comment (jfhbrook) [inline pyee/_base.py:143]: I'd be curious to see if this shows up in the sphinx docs - the makefile has tasks for this if you wanna give it a shot. If it shows up I'll want to remove this docstring, but otherwise this seems ok to me - though I will say, the word naked makes me giggle a little.

**Review by jfhbrook — CHANGES_REQUESTED** [review]
This loops pretty good! I think beyond stylistic things what I'd like to see implementation-wise is the lock used in the other non-removal spots where things are mucked around with, particularly in `_add_event_handler` and `_call_handlers`.

It would also be great if we can add a test for the sync handler, something like:

```py
ee = EventEmitter()

@ee.once("event")
def remove_self():
    ee.remove_listener("event", remove_self)

ee.emit("event")
```

This should crash before this change but work afterwards.

Thanks for taking a look at this! I really appreciate it, it's saving me a lot of work.

**Comment by forslund** [comment]
Good comments all around will try to get an update done tonight / tomorrow

**Comment by forslund** [comment]
Good comments all around will try to get an update done tonight / tomorrow

**Comment by jfhbrook** [comment]
Cool! Any work you do on this is much appreciated, and if you run out of steam I can always take and run with your work.

File: pyee/_base.py (hunk 1 above)
Comment (forslund) [reply pyee/_base.py:143]: I've verified that the this docstring doesn't get pulled into the sphinx docs. It seems Sphinx ignores all "private" methods / functions unless `:private-members:` is specified for the class

Comment (jfhbrook) [reply pyee/_base.py:143]: Nice!

File: pyee/_base.py [hunk 2]
```diff
@@ -38,6 +39,7 @@ def on_error(message):
     """
     def __init__(self):
         self._events = defaultdict(OrderedDict)
+        self.lock = Lock()
```
[the remark this answers is not in the document]
Comment (forslund) [reply pyee/_base.py]: Changed it to `_lock`

**Comment by forslund** [comment]
The testcase became a bit more convoluted since the removal is done before the target function is called (I guess to minimize the risk of this kind of race). Let me know if the solution is too hacky.

File: pyee/_base.py [hunk 3]
```diff
@@ -118,7 +123,13 @@ def once(self, event, f=None):
...
         def _wrapper(f):
             def g(*args, **kwargs):
-                self.remove_listener(event, f)
+                with self._lock:
+                    # Check that the event wasn't removed already right
+                    # before the lock
+                    if event in self._events and f in self._events[event]:
+                        self._remove_listener(event, f)
+                    else:
+                        return None
```
Comment (jfhbrook) [inline pyee/_base.py:132]: Yeah OK I see what you mean here. Interesting, I wouldn't have expected the edge case we saw given this ordering (though in a threaded context without locks it makes sense). Guess I should have looked more closely!

Mulling this over, I *think* shrugging and returning None is OK, because if the edge case we're talking about triggers it probably means that the handler has either already been called or was  intentionally removed between the function executing and the lock yielding.

File: tests/test_sync.py [hunk 4]
```diff
@@ -203,6 +204,21 @@ def once_handler(data):
     assert ee._events['event'] == OrderedDict()
 
 
+def test_once_remove_race():
```
Comment (jfhbrook) [inline tests/test_sync.py:207]: Oof, yeah, OK, I see what you mean with this test. I wish there was a sensible way to test this, but I think that given what we have here it's too hairy to implement what we had in mind. My vote is to remove this test and shrug.

**Review by jfhbrook — CHANGES_REQUESTED** [review]
I think we can remove the test - you're right that's really hairy - and then I'm happy to merge this. Nice work!

**Comment by forslund** [comment]
I dropped the test commit from the branch and squashed the remaining two commits into a single one. should be good to go now.

**Review by jfhbrook — APPROVED** [review]
Excellent. I'll try to do a merge and release tonight! This weekend at the latest.

**Comment by jfhbrook** [comment]
Released in 8.1.0. Thanks!
